x/ibc/07-tendermint/types/proposal_handle_test.go

```diff
@@ -0,0 +1,256 @@
+package types_test
+
+import (
+	clienttypes "github.com/cosmos/cosmos-sdk/x/ibc/02-client/types"
+	"github.com/cosmos/cosmos-sdk/x/ibc/07-tendermint/types"
+	"github.com/cosmos/cosmos-sdk/x/ibc/exported"
+)
+
+var (
+	frozenHeight = clienttypes.NewHeight(0, 1)
+)
+
+// sanity checks
+func (suite *TendermintTestSuite) TestCheckProposedHeaderAndUpdateStateBasic() {
+	clientA, _ := suite.coordinator.SetupClients(suite.chainA, suite.chainB, exported.Tendermint)
+	clientState := suite.chainA.GetClientState(clientA).(*types.ClientState)
+	clientStore := suite.chainA.App.IBCKeeper.ClientKeeper.ClientStore(suite.chainA.GetContext(), clientA)
+
+	// use nil header
+	cs, consState, err := clientState.CheckProposedHeaderAndUpdateState(suite.chainA.GetContext(), suite.chainA.App.AppCodec(), clientStore, nil)
+	suite.Require().Error(err)
+	suite.Require().Nil(cs)
+	suite.Require().Nil(consState)
+
+	clientState.LatestHeight = clientState.LatestHeight.Increment()
+
+	// consensus state for latest height does not exist
+	cs, consState, err = clientState.CheckProposedHeaderAndUpdateState(suite.chainA.GetContext(), suite.chainA.App.AppCodec(), clientStore, suite.chainA.LastHeader)
+	suite.Require().Error(err)
+	suite.Require().Nil(cs)
+	suite.Require().Nil(consState)
+
+}
+
+// to expire clients, time needs to be fast forwarded on both chainA and chainB.
+// this is to prevent headers from failing when attempting to update later.
+func (suite *TendermintTestSuite) TestCheckProposedHeaderAndUpdateState() {
```

Comment:
The structure of this test is confusing and difficult to understand imo. I suggest a clearer approach